tools/hsmtool.c

```diff
@@ -368,6 +371,145 @@ static int guess_to_remote(const char *address, struct node_id *node_id,
+	printf("  4) Japanese (jp)\n");
+	printf("  5) Chinese simplified (zhs)\n");
+	printf("  6) Chinese traditional (zht)\n");
+	printf("Select [0-6]: ");
+
+	char *selection=NULL;
+	size_t size = 0;
+	size_t characters = getline(&selection, &size, stdin);
+	if (characters < 0)
+		errx(ERROR_USAGE, "Could not read line from stdin.");
+	int selected = atoi(selection);
+
+	char *lang = malloc(4);
+	switch (selected) {
+		case 0:
+			strcpy(lang, "en");
+			break;
+		case 1:
+			strcpy(lang, "es");
+			break;
+		case 2:
+			strcpy(lang, "fr");
+			break;
+		case 3:
+			strcpy(lang, "it");
+			break;
+		case 4:
+			strcpy(lang, "jp");
+			break;
+		case 5:
+			strcpy(lang, "zhs");
+			break;
+		case 6:
+			strcpy(lang, "zht");
+			break;
+		default:
+			errx(ERROR_USAGE, "Invalid language selection, select one from the list [0-6].");
+	}
+
+	bip39_get_wordlist(lang, words);
+}
+
+static void get_mnemonic(char *mnemonic) {
+	char *line = NULL;
+	size_t line_size = 0;
+
+	printf("Introduce your BIP39 word list:\n");
+	size_t characters = getline(&line, &line_size, stdin);
+	if (characters < 0)
+		errx(ERROR_USAGE, "Could not read line from stdin.");
+	line[characters-1] = '\0';
+	strcpy(mnemonic, line);
+	free(line);
+}
+
+static void read_mnemonic(char *mnemonic) {
+	/* Get words for the mnemonic language */
+	struct words *words;
+	get_words(&words);
+
+	/* Get mnemonic */
+	get_mnemonic(mnemonic);
+
+	if (bip39_mnemonic_validate(words, mnemonic) != 0) {
+		errx(ERROR_USAGE, "Invalid mnemonic: \"%s\"", mnemonic);
+	}
+}
+
+static void read_passphrase(char **passphrase) {
+	struct termios current_term, temp_term;
+	printf("Warning: remember that different passphrases yield different "
+	       "bitcoin wallets.\n");
+	printf("Leave it empty for using the value by default (echo is "
+	       "disabled now).\n");
+	printf("Enter your passphrase: \n");
+
+	/* Change terminal options so we do not echo the passphrase */
+	if (tcgetattr(fileno(stdin), &current_term) != 0)
+		errx(ERROR_USAGE, "Could not get current terminal options.");
+	temp_term = current_term;
+	temp_term.c_lflag &= ~ECHO;
+	if (tcsetattr(fileno(stdin), TCSAFLUSH, &temp_term) != 0)
+		errx(ERROR_USAGE, "Could not disable passphrase echoing.");
+	/* If we don't flush we might end up being buffered and we might seem
+	 * to hang while we wait for the password. */
+	fflush(stdout);
+
+	size_t passphrase_size = 0;
+	size_t characters = getline(passphrase, &passphrase_size, stdin);
+	if (characters < 0)
+		errx(ERROR_USAGE, "Could not read passphrase from stdin.");
+
+	/* Replace the last character because we assume it was the newline '\n' */
+	(*passphrase)[characters-1] = '\0';
+
+	if (tcsetattr(fileno(stdin), TCSAFLUSH, &current_term) != 0)
+		errx(ERROR_USAGE, "Could not restore terminal options.");
+}
+
+static int generate_hsm(const char *hsm_secret_path)
+{
+	char mnemonic[BIP39_WORDLIST_LEN];
+	read_mnemonic(mnemonic);
+
+	char *passphrase = NULL;
+	read_passphrase(&passphrase);
```

Comment:
If `read_passphrase` set `passphrase`, it's never freed

Reply:
Thanks for this,  is there any automatic tool that checks for this silly cases that I forgot? (so I can use it before future contributions and you do not have to lose your time on "simple" memory leaks)

Reply:
Just went looking for missing `free`, but you can test with valgrind or clang with leak sanitizer enabled (`configure --enable-address-sanitizer`) to at least make sure the frequent paths don't leak.

Reply:
Even better: [write a test](https://github.com/ElementsProject/lightning/blob/d151d554b37e8cbf117d3a85f1503588830da8a8/tests/test_wallet.py#L988) to trigger those codepaths :)